rfcs/111-add-service-metadata-in-storage-metadata.md

```diff
@@ -0,0 +1,86 @@
+---
+author: JinnyYi <github.com/JinnyYi>
+status: draft
+updated_at: 2021-06-17
+---
+
+# GSP-111: Add Service Metadata In Storage Metadata
+
+## Background
+
+In [GSP-6], we split storage meta and object meta. In [GSP-40], we split all object metadata into four groups: `global` metadata, `standard` metadata, `service` metadata and `user` metadata. For storage, there's no `standard` metadata and user defined metadata is not allowed.
+
+For now, storage related information carried in `StorageMeta`:
+
+```go
+type StorageMeta struct {
+	location string
+	Name     string
+	WorkDir  string
+
+	// bit used as a bitmap for object value, 0 means not set, 1 means set
+	bit uint64
+	m   map[string]interface{}
+}
+```
+
+It only defines the `global` metadata for storage which could be used in all services, but no field to carry the metadata used in special services.
+
+## Proposal
+
+So I propose to add service metadata in storage metadata.
+
+`service` metadata is provided by service, could only be used in current service.
+
+The `global` metadata has been included in object type. `service` metadata will be stored in a struct that defined by service:
+
+- service metadata will be `serviceMetadata interface{}` in `StorageMeta`.
+
+For services side:
+
+Example of adding service pair for storage metadata in `service.toml`:
+
+```go
+[infos.storage.meta.<service-meta>]
+type = "<type>"
+description = "<description>"
+```
+
+For service metadata, we will introduce `Strong Typed Service Metadata` like `ObjectMetadata`. We will generate following struct for specific service according to service pairs:
+
+```go
+type ServiceMetadata struct {
+    <service meta>  <type>
+    ...
+}
+```

```

Comment:
Do I understand correctly?
- In `go-storage/types`, there are type `Object` (has field `serviceMetadata interface{}`), and `StorageMeta`
- In `go-service-*` type `ObjectMetadata` is used to set `serviceMetadata`, and has getter `GetObjectMetadata`

Now, this proposal intends to also add the field `serviceMetadata interface{}` to  `StorageMeta`. Should the `go-service-*` type be called `StorageMetadata`?

And I'm wondering why type `Object` and `StorageMeta` are at the same level :thinking:

Reply:
Do we need to rename `go-service-*.ObjectMetadata` into something like `go-service-*.ServiceObjectMetadata` to avoid confusion? (and correspondingly, `ServiceStorageMeta` in this proposal.)

Reply:
The root cause is `Object` is a struct but `Storager` is an interface.

If `Storage` is a struct that exposed to the user, we can set `serviceMetadata` in `Storage` struct just like we do on `Object`.

Well, `StorageMeta` and `StorageMetadata` do look wired to me too. Do you have a better idea/name for that?

Reply:
I think the most wired one is that both `go-storage.Object` and `go-storage.StorageMeta` will have field `serviceMetadata`, 


but `go-service-*.ServiceMetadata` will be the one that belongs to `StorageMeta`. This seems to be rather confusing

Reply:
I got it.

The problem is introduced by simplify

- `Service Level/Specified Object Metadata` to `serviceMetadata`
- `Service Level/Specified Storage Metadata` to `serviceMetadata`

But I don't have better ideas about the name.

Reply:
What about `ServiceObjectMetadata` and `ServiceStorageMetadata`?

Reply:
Maybe we can find the opposite word of `global` to represent all those situations? So we can:

- Distinguish `Service Pair` and `Service's Pair`
- Distinguish `Service Metadata` and `Service's Metadata`
- ...

And our API will turn into:

```go
func GetObjectXxxxMetadata(o* Object) ObjectXxxxMetadata
func GetStorageXxxxMetadata(s *StorageMeta) StorageXxxxMetadata
```

Reply:
> - Distinguish `Service Pair` and `Service's Pair`
> - Distinguish `Service Metadata` and `Service's Metadata`

Do I understand correctly?
- `Service Pair` could be all the pairs in the current service. `Service's Pair` just means pairs related to the service implementations. 
- `Service Metadata` includes `object service metadata` and `storage service metadata`.  `Service's Metadata` could corresponds to `StorageMeta` here.